### PR TITLE
Prediction output reification

### PR DESCRIPTION
**Pull request by edobobo**

### CODE REVIEW

File: classy/data/data_modules.py [hunk 1]
```diff
@@ -201,15 +201,19 @@ def prepare_data(self) -> None:
         elif self.external_vocabulary_path is not None:
             logger.info("Loading vocabulary from external passed directory")
             self.vocabulary = Vocabulary.from_folder(self.external_vocabulary_path)
-            self.vocabulary.save("vocabulary")
```
Comment (Valahaar) [inline classy/data/data_modules.py:204]: why has this been changed?

File: classy/evaluation/base.py [hunk 2]
```diff
@@ -6,11 +6,6 @@
 class Evaluation:
     def __call__(
         self,
-        predicted_samples: List[
-            Tuple[
-                Union[SentencePairSample, SequenceSample, TokensSample, QASample, GenerationSample],
-                Union[str, List[str], Tuple[int, int]],
-            ]
-        ],
+        predicted_samples: List[Union[SentencePairSample, SequenceSample, TokensSample, QASample, GenerationSample]],
```
Comment (Valahaar) [inline classy/evaluation/base.py:9]: I'm not sure if we should create a grouping typing variable ClassificationSample that is the union of SentencePairSample, SequenceSample, TokensSample, QASample :thinking:

File: classy/evaluation/simple.py [hunk 3]
```diff
@@ -21,25 +21,23 @@ def p_r_f_support(gold, pred) -> Dict[str, float]:
 
 
 class SequenceSimpleEvaluation(Evaluation):
```
Comment (Valahaar) [inline classy/evaluation/simple.py:23]: are these all the same now (aside from typing, which we can Union-ize)? should we merge them?

File: docs/docs/getting-started/customizing-things/custom-dataset.md [hunk 4]
```diff
@@ -57,12 +57,13 @@ You first deal with the vocabulary methods. As you are doing sequence classifica
 
 ```python
     @staticmethod
```
Comment (Valahaar) [inline docs/docs/getting-started/customizing-things/custom-dataset.md:59]: does this need dedention?

File: docs/docs/getting-started/customizing-things/custom-dataset.md [hunk 5]
```diff
@@ -114,18 +115,18 @@ Finally, you need to implement the *dataset_iterator_func*:
 
 ```python
     def dataset_iterator_func(self) -> Iterable[Dict[str, Any]]:
```
Comment (Valahaar) [inline docs/docs/getting-started/customizing-things/custom-dataset.md:117]: same here?

File: classy/data/data_drivers.py [hunk 6]
```diff
@@ -345,35 +419,41 @@ def read(self, lines: Iterator[str]) -> Generator[TokensSample, None, None]:
...
             for sample in samples:
-                f.write("\t".join([" ".join(sample.tokens), " ".join(sample.labels)]) + "\n")
+                used_annotation = (
+                    sample.predicted_annotation if use_predicted_annotation else sample.reference_annotation
+                )
+                f.write("\t".join([" ".join(sample.tokens), " ".join(used_annotation)]) + "\n")
 
 
 class JSONLTokensDataDriver(TokensDataDriver):
     def read(self, lines: Iterator[str]) -> Generator[TokensSample, None, None]:
```
Comment (Valahaar) [inline classy/data/data_drivers.py:432]: Iterator

File: classy/data/data_drivers.py [hunk 7]
```diff
@@ -445,29 +535,38 @@ def read(self, lines: Iterator[str]) -> Generator[GenerationSample, None, None]:
...
+                    sample.predicted_annotation if use_predicted_annotation else sample.reference_annotation
+                )
+                if used_annotation is None:
+                    f.write(f"{sample.source_sequence}\n")
+                else:
+                    f.write(f"{sample.source_sequence}\t{used_annotation}\n")
 
 
 class JSONLGenerationDataDriver(GenerationDataDriver):
     def read(self, lines: Iterator[str]) -> Generator[GenerationSample, None, None]:
```
Comment (Valahaar) [inline classy/data/data_drivers.py:554]: Iterator

File: classy/pl_modules/mixins/saving.py [hunk 8]
```diff
@@ -22,10 +22,7 @@ def save_resources_and_update_config(
         # save examples
         source, examples = data_module.get_examples(n=5)
         experiment_folder.joinpath("data").mkdir(exist_ok=True)
-        get_data_driver(self.task, "jsonl").save(
-            examples,
-            str(experiment_folder / "data" / f"examples-{source}.jsonl"),
-        )
+        get_data_driver(self.task, "jsonl").save(examples, str(experiment_folder / "data" / f"examples-{source}.jsonl"))
```
Comment (Valahaar) [inline classy/pl_modules/mixins/saving.py:25]: it seems correct to me, but double checking: here we want the gold label, right?

File: classy/evaluation/base.py (hunk 2 above)
Comment (poccio) [reply classy/evaluation/base.py:9]: Actually, why aren't we using ClassyStruct?

Comment (Valahaar) [reply classy/evaluation/base.py:9]: if we only use ClassyStruct to subclass samples, I'd use it and rename it to ClassySample maybe

File: classy/pl_modules/mixins/saving.py (hunk 8 above)
Comment (poccio) [reply classy/pl_modules/mixins/saving.py:25]: Yup

File: classy/evaluation/simple.py (hunk 3 above)
Comment (edobobo) [reply classy/evaluation/simple.py:23]: Actually I don't think so, they can change in the future it is more clear to me if we keep them separated.

File: classy/evaluation/base.py (hunk 2 above)
Comment (edobobo) [reply classy/evaluation/base.py:9]: I agree with that, we should use ClassySammple in imo. I am changing it

**Review by Valahaar — APPROVED** [review]
LGTM! can merge

**Pull request merged**